Schnorr signature

gtillem opened the pull request:
PR for the implementation of Schnorr signatures. The code implements a variant of Schnorr signatures, EdDSA on AltBabyJubJub curve using Blake2 hash functions. 

The signing algorithm has two variants: One with the raw message, i.e.,  the message is not hashed before signing, and one with message hashing.

gtillem committed:
Add fixed generators that are used as base points

gtillem committed:
Implement schnorr signatures

gtillem committed:
Test schnorr signatures

gtillem committed:
Comment out benchmark tests

vnermolaev commented on src/main/kotlin/com/ing/dlt/zkkrypto/ecc/schnorr/FixedGenerators.kt:
```diff
@@ -0,0 +1,54 @@
+    /**
+     * Base points to be used in different protocols. Right now we only use SpendingKeyGenerator for Schnorr signatures
+     */
+    fun altBabyJubjubFixedGenerators(): List<EllipticCurvePoint> {
```
try
```
val altBabyJubjubFixedGenerators = listOf(...)
```

vnermolaev commented on src/main/kotlin/com/ing/dlt/zkkrypto/ecc/schnorr/SchnorrSignature.kt:
```diff
@@ -0,0 +1,162 @@
+import java.security.SecureRandom
+import kotlin.random.Random.Default.nextBytes
+
+enum class Hash{
```
is this not already defined somewhere in the lib? because Alexei did one of this hashes.

gtillem replied:
It's not, afaik. But I removed this outside of this class and defined as an enum with other supported hash functions. Because, in the future, Mimc or Poseidon can be used here.

vnermolaev commented on src/main/kotlin/com/ing/dlt/zkkrypto/ecc/schnorr/SchnorrSignature.kt:
```diff
@@ -0,0 +1,162 @@
+    val msgPersonalization: ByteArray) {
+
+    companion object {
+        fun zinc() = SchnorrSignature(
```
try `val` instead of `fun`

vnermolaev commented on src/main/kotlin/com/ing/dlt/zkkrypto/ecc/schnorr/SchnorrSignature.kt:
```diff
@@ -0,0 +1,162 @@
+        const val BLAKE2S_DIGEST_LENGTH = 32
+    }
+
+    private lateinit var privateKey: BigInteger
```
This is arguable. Consider this
```
class SchnorrSignature {
  ...
  val privateKey: BigInteger
  val publicKey: BigInteger
   
  init {
    nextKeyPair()
  }

  fun nextKeyPair() {
    privateKey = BigInteger(curve.S.bitLength(), SecureRandom())
        while (privateKey > curve.S) {
            privateKey = BigInteger(curve.S.bitLength(), SecureRandom())
        }
      publicKey = base.scalarMult(privateKey)
  }
   ...
}
```
then you won't need to do 
```
schnorr.generatePrivateKey()
        schnorr.getPublicKeyFromPrivate()
```
after each instantiation and can call `nextKeyPair` whenever needed (as you do in a loop in one of the tests)

vnermolaev commented on src/main/kotlin/com/ing/dlt/zkkrypto/ecc/schnorr/SchnorrSignature.kt:
```diff
@@ -0,0 +1,162 @@
+        val uniformMessage = toUniformFieldElement(paddedMessage)
+
+        //second component of the signature
+        val s = uniformMessage.multiply(privateKey).mod(curve.S).add(r).mod(curve.S)
```
optionally can be written with `+`, `*` and `%`

vnermolaev commented on src/main/kotlin/com/ing/dlt/zkkrypto/ecc/schnorr/SchnorrSignature.kt:
```diff
@@ -0,0 +1,162 @@
+        val t = nextBytes(80)
+
+        //Generate random r as the hash digest of Blake2b(t || msg)
+        val r = hashToScalar(Hash.BLAKE2B, compPersonalization, first = t, second = msgBytes)
```
optionally, `t` can be inlined

vnermolaev commented on src/main/kotlin/com/ing/dlt/zkkrypto/ecc/schnorr/SchnorrSignature.kt:
```diff
@@ -0,0 +1,162 @@
+        return Signature(rPoint, s)
+    }
+
+    fun signHashedMessage(msgBytes: ByteArray) : Signature {
```
see comments to signRawMessage

vnermolaev commented on src/main/kotlin/com/ing/dlt/zkkrypto/ecc/schnorr/SchnorrSignature.kt:
```diff
@@ -0,0 +1,162 @@
+    private fun hashToScalar(hashFunction: Hash, personalization: ByteArray, first: ByteArray, second:ByteArray) : BigInteger {
+        val digest = when (hashFunction) {
+            Hash.BLAKE2B -> hashBlake2b(first + second, personalization);
+            Hash.BLAKE2S -> hashBlake2s(first.plus(second), personalization)
```
please align both arms

vnermolaev commented on src/main/kotlin/com/ing/dlt/zkkrypto/ecc/schnorr/SchnorrSignature.kt:
```diff
@@ -0,0 +1,162 @@
+            val ubyte = byte.asUnsigned()
+            (7 downTo 0).fold(acc) { local, shift ->
+                val bit = ubyte shr shift and 1
+                (local + local) % curve.S + if (bit == 1) { BigInteger.ONE } else { BigInteger.ZERO }
```
now that I see there is `scalarMult`, it is possible to rewrite as
`local.scalarMult(2) % curve.S + BigInteger.ONE.scalarMult(bit)`

gtillem replied:
`scalarMult()` is an EllipticCurvePoint function. It performs the scalar multiplication on the point, not on the BigIntegers. So, I can't use it in here. But alternatively, I can use one of the following:

`(local + local) % curve.S + BigInteger.valueOf(bit.toLong())`

`(local + local) % curve.S + BigInteger.(bit.toString())`

vnermolaev commented on src/main/kotlin/com/ing/dlt/zkkrypto/ecc/schnorr/Signature.kt:
```diff
@@ -0,0 +1,8 @@
+import com.ing.dlt.zkkrypto.ecc.EllipticCurvePoint
+import java.math.BigInteger
+
+data class Signature(val r:EllipticCurvePoint, val s: BigInteger) {
```
remove extra `{ }`

vnermolaev commented on src/test/kotlin/com/ing/dlt/zkkrypto/ecc/schnorr/SchnorrTest.kt:
```diff
@@ -0,0 +1,180 @@
+        val msg = 85.toByte()
+        val msgBytes = byteArrayOf(msg)
+
+        //val msg = "Foo bar pad to16"
```
remove commented code or make two assertions in the test, one for 85 and one more ByteArray, for example
```
listOf(ByteArray(1){ 85 }, "Foo bar pad to16".toByteArray).forEach {
  ... verification logic
}
```

vnermolaev commented on src/test/kotlin/com/ing/dlt/zkkrypto/ecc/schnorr/SchnorrTest.kt:
```diff
@@ -0,0 +1,180 @@
+        println("\tx coord = " + schnorr.publicKey.x.toString(16))
+        println("\ty coord = " + schnorr.publicKey.y.toString(16) + "\n")
+
+        val msg = 85.toByte()
```
see comment in the prev test

vnermolaev commented on src/test/kotlin/com/ing/dlt/zkkrypto/ecc/schnorr/SchnorrTest.kt:
```diff
@@ -0,0 +1,180 @@
+        assert(verified)
+    }
+
+    //@Test
```
Either remove the comment or write `@Test @Ignore`

mvdbos replied:
Or better, upgrade to Junit5 as I did for ZKNotary, and tag these tests with 'benchmark' and create a 'benchmark' Gradle task, so that they can be run separately.

vnermolaev commented on src/test/kotlin/com/ing/dlt/zkkrypto/ecc/schnorr/SchnorrTest.kt:
```diff
@@ -0,0 +1,180 @@
+            finishKeyGen += System.nanoTime()
+
+            //measure sign raw message
+            startSign += System.nanoTime()
```
Consider https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.system/measure-nano-time.html

mvdbos replied:
Agreed. Especially in test code, it is ok to use experimental.

vnermolaev commented:
Please see `spotlessApply` task in `zk-notary` to format your code.

In general, very usefu, looking forward to embedd this in into the `zk-notary`!

mvdbos commented on src/main/kotlin/com/ing/dlt/zkkrypto/ecc/schnorr/SchnorrSignature.kt:
```diff
@@ -0,0 +1,162 @@
+        val rPoint = base.scalarMult(r)
+
+        //pad message
+        val paddedMessage = msgBytes + ByteArray(MAX_MESSAGE_SIZE - msgBytes.size)
```
What if the msgBytes.size is larger than MAX_MESSAGE_SIZE? Shouldn't there be some kind `require` or `check`?

mvdbos replied:
NegativeArraySizeException...

mvdbos commented on src/main/kotlin/com/ing/dlt/zkkrypto/ecc/schnorr/SchnorrSignature.kt:
```diff
@@ -0,0 +1,162 @@
+
+    fun verifyRawMessage(msgBytes: ByteArray, signature: Signature) : Boolean {
+        //pad message to max message size and convert to a field element
+        val paddedMessage = msgBytes + ByteArray(MAX_MESSAGE_SIZE - msgBytes.size)
```
Same here as on `signRawMessage`: add checks on length?

mvdbos reviewed (changes requested):
First of all: nice work! really cool to have this. 

I have only two minor comments for now.

One question: I see some of Victor's comments marked as resolved, but no code changed for it. Is that the result of agreement between you outside GitHub comments? I have +1'd the ones I feel still need to be addressed.

Finally, I agree with Victor that this would be a good time to 'upgrade' the build of this project a bit with both Junit5 and Spotless. Alternatively do it directly after in a separate PR. It is easily copied from the ZK Notary project. I can help with that.

gtillem committed:
Update signature protocol

gtillem committed:
Update benchmark tests

gtillem committed:
Add spotlessApply

mvdbos commented on src/main/kotlin/com/ing/dlt/zkkrypto/ecc/schnorr/SchnorrSignature.kt:
```diff
@@ -40,63 +38,71 @@ class SchnorrSignature (
-    fun generatePrivateKey() {
+    private var privateKey: BigInteger
+    var publicKey: EllipticCurvePoint
+    init {
```
I am not yet happy with this.

1) I like Victor's proposal from yesterday better, because it doesn't use `var` (even though that means his proposal doesn't work). The class should be immutable in my opinion. So taht means nextKeyPair should just return a new instance. 
2) SecureRandom should be part of the constructor, so we can change it, especially for tests.
3) This class now has two responsibilities: signing/verifying and keypair generation. We can solve 1) and 2) by extracting the keypair generation out and making the keys arguments to the sign and verify methods. This is also more in line with other signign APIs.

mvdbos commented on src/main/kotlin/com/ing/dlt/zkkrypto/ecc/schnorr/SchnorrSignature.kt:
```diff
@@ -40,63 +38,71 @@ class SchnorrSignature (
-        //Generate random r as the hash digest of Blake2b(t || msg)
-        val r = hashToScalar(Hash.BLAKE2B, compPersonalization, first = t, second = msgBytes)
+        // Generate random r as the hash digest of Blake2b(randomBytes || msg)
+        val r = hashToScalar(HashEnum.BLAKE2B, compPersonalization, first = nextBytes(80), second = msgBytes)
```
Question out of ignorance: why do we use BLAKE2B and 2S both? Why not only 2S? Or is that part of a spec?

mvdbos commented on src/main/kotlin/com/ing/dlt/zkkrypto/ecc/schnorr/SchnorrSignature.kt:
```diff
@@ -40,63 +38,71 @@ class SchnorrSignature (
     }
 
-    fun signHashedMessage(msgBytes: ByteArray) : Signature {
+    fun signHashedMessage(msgBytes: ByteArray): Signature {
```
This now also only supports messages of MAX_MESSAGE_SIZE. What is the added value above the raw version then? Why don't we make it support any length? That should be possible with Blake2s, right?

mvdbos commented on src/test/kotlin/com/ing/dlt/zkkrypto/ecc/schnorr/SchnorrTest.kt:
```diff
@@ -1,180 +1,151 @@
-
+import kotlin.system.measureNanoTime
 
 class SchnorrSignatureTest {
```
I would like to see more tests (preferably one parametrized test) that use test vectors that prove that this is really a conforming implementation. Are there test vectors we can get from somewhere? Possibly not for this specific curve, but I think vectors exist for Secp256k1?

gtillem replied:
I added tests with test vectors. I used the message values from Zinc and from Corda EdDSA, which are smaller than 32 bytes. I also needed to add a setter function for keys and remove seed generation out of the signing function to be able to compare the generated values with the test vectors.

mvdbos commented on src/main/kotlin/com/ing/dlt/zkkrypto/ecc/schnorr/SchnorrSignature.kt:
```diff
@@ -2,35 +2,33 @@ package com.ing.dlt.zkkrypto.ecc.schnorr
  */
 
-class SchnorrSignature (
+class SchnorrSignature(
```
Have you already looked at how to integrate this into Corda through/as a Security Provider? I thought to be able to do that, the `Signature` implementation has to be a `SignatureSpi`: 
```
 * This class defines the <i>Service Provider Interface</i> (<b>SPI</b>)
 * for the {@code Signature} class, which is used to provide the
 * functionality of a digital signature algorithm. Digital signatures are used
 * for authentication and integrity assurance of digital data.
 *.
 * <p> All the abstract methods in this class must be implemented by each
 * cryptographic service provider who wishes to supply the implementation
 * of a particular signature algorithm.
```
And this would also require you to make the keys as standard keys: `java.security.PublicKey` and `java.security.PrivateKey`.

That said, I am not a specialist in this area, I just looked at where Corda gets its supported signature schemes from and how they are used. Perhaps you have found another, easier way in which the current API you designed can be used.

mvdbos committed:
Introduce benchmark Gradle task

mvdbos committed:
Remove inline function warning

Warning: 'Expected performance impact from inlining is insignificant.
Inlining works best for functions with parameters of functional types'.

So removed inline prefix.

gtillem committed:
Update functions to enable reading seed and keys for test vectors

gtillem committed:
Add test vectors

gtillem committed:
Reformat code

alexeykoren commented:
Maybe also update readme file?

gtillem committed:
Update readme file

gtillem commented:
> Maybe also update readme file?

Updated it now!
